File: apps/mobile/ios/HermesMobile/State/SessionLibrary.swift
import Foundation

enum SessionLibraryView: String, CaseIterable, Identifiable, Sendable {
    case active
    case archived
    var id: String { rawValue }
}

enum SessionLibrary {
    static func filter(_ sessions: [SessionSummary], view: SessionLibraryView, query: String) -> [SessionSummary] {
        let needle = query.trimmingCharacters(in: .whitespacesAndNewlines)
        return sessions.filter { session in
            let belongs = session.archived == (view == .archived)
            return belongs && (needle.isEmpty || matches(session, needle))
        }
    }

    private static func matches(_ session: SessionSummary, _ query: String) -> Bool {
        let fields = [session.displayTitle, session.title, session.preview, session.storedID]
        return fields.contains { $0.range(of: query, options: [.caseInsensitive, .diacriticInsensitive]) != nil }
    }
}

File: apps/mobile/android/app/src/main/java/com/snowzlmbot/hermes/mobile/feature/SessionLibrary.kt
package com.snowzlmbot.hermes.mobile.feature

import com.snowzlmbot.hermes.mobile.core.SessionSummary
import java.util.Locale

internal enum class SessionLibraryView { ACTIVE, ARCHIVED }

internal object SessionLibrary {
  fun filter(
    sessions: List<SessionSummary>,
    view: SessionLibraryView,
    query: String,
  ): List<SessionSummary> {
    val normalized = query.trim().lowercase(Locale.ROOT)
    return sessions.filter { session ->
      val inView = session.archived == (view == SessionLibraryView.ARCHIVED)
      inView && (
        normalized.isEmpty() ||
          session.displayTitle.matches(normalized) ||
          session.title.matches(normalized) ||
          session.preview.matches(normalized) ||
          session.storedId.matches(normalized)
      )
    }
  }

  private fun String.matches(query: String): Boolean = lowercase(Locale.ROOT).contains(query)
}
